src/ucp/core/ucp_mm.c

```diff
@@ -554,44 +637,77 @@ ucs_status_t ucp_mem_map(ucp_context_h context, const ucp_mem_map_params_t *para
         goto out;
     }
 
+    length = params->length;
+
     if (flags & UCP_MEM_MAP_ALLOCATE) {
-        memory_type = UCP_PARAM_VALUE(MEM_MAP, params, memory_type, MEMORY_TYPE,
-                                      UCS_MEMORY_TYPE_HOST);
+        mem_type = UCP_PARAM_VALUE(MEM_MAP, params, memory_type, MEMORY_TYPE,
+                                   UCS_MEMORY_TYPE_HOST);
     } else if (!(params->field_mask & UCP_MEM_MAP_PARAM_FIELD_MEMORY_TYPE) ||
                (params->memory_type == UCS_MEMORY_TYPE_UNKNOWN)) {
-        ucp_memory_detect(context, address, params->length, &mem_info);
-        memory_type = mem_info.type;
+        ucp_memory_detect(context, address, length, &mem_info);
+        mem_type = mem_info.type;
     } else {
         if (params->memory_type > UCS_MEMORY_TYPE_LAST) {
             ucs_error("invalid memory type %d", params->memory_type);
             status = UCS_ERR_INVALID_PARAM;
             goto out;
         }
 
-        memory_type = params->memory_type;
+        mem_type = params->memory_type;
+    }
+
+    uct_flags = UCT_MD_MEM_ACCESS_ALL;
+
+    if (params->field_mask & UCP_MEM_MAP_PARAM_FIELD_FLAGS) {
+        if (params->flags & UCP_MEM_MAP_NONBLOCK) {
+            uct_flags |= UCT_MD_MEM_FLAG_NONBLOCK;
+        }
+
+        if (params->flags & UCP_MEM_MAP_FIXED) {
+            uct_flags |= UCT_MD_MEM_FLAG_FIXED;
+        }
+    }
+
+    reg_md_map = context->reg_md_map[mem_type];
+
+    status = ucp_memh_create_no_rcache(context, address, length, mem_type,
+                                       &memh);
+    if (status != UCS_OK) {
+        goto out;
     }
 
     if (flags & UCP_MEM_MAP_ALLOCATE) {
-        status = ucp_memh_alloc(context, address, params->length, memory_type,
-                                ucp_mem_map_params2uct_flags(params),
-                                "user memory", memh_p);
-    } else {
-        status = ucp_memh_get(context, address, params->length, memory_type,
-                              context->reg_md_map[memory_type],
-                              ucp_mem_map_params2uct_flags(params), memh_p);
+        status = ucp_mem_do_alloc(context, address, length, uct_flags, mem_type,
+                                  &mem);
+        if (status != UCS_OK) {
+            goto err_free_memh;
+        }
+
+        ucp_memh_init_from_mem(context, memh, &mem);
+        reg_md_map &= ~memh->md_map;
```

Comment:
i would expect reg_md_map to always become 0 here

Reply:
Why? should be all except alloc_md

Reply:
AFAIU, ucp_mem_do_alloc should already register all of `context->reg_md_map[mem_type]`